src/AWS/Lambda/Events/ApiGateway/ProxyRequest.hs

```diff
@@ -0,0 +1,167 @@
+--
+-- The 'ProxyRequest' notably has one parameter for the type of information
+-- returned by the API Gateway's custom authorizer (if applicable).  This type
+-- must also implement FromJSON so that it can be decoded.  If you do not
+-- expect this data to be populated we recommended using the 'NoAuthorizer'
+-- type exported from this module (which is just an alias for 'Value').  If
+-- there _must not_ be authorizer populated (this is unlikely) then use the
+-- 'StrictlyNoAuthorizer' type.
+--
+-- @
+--     {-\# LANGUAGE NamedFieldPuns \#-}
+--     {-\# LANGUAGE DuplicateRecordFields \#-}
+--
+--     module Main where
+--
+--     import AWS.Lambda.Runtime (pureRuntime)
+--     import AWS.Lambda.Events.ApiGateway.ProxyRequest (ProxyRequest(..), NoAuthorizer)
+--     import AWS.Lambda.Events.ApiGateway.ProxyResponse (ProxyResponse(..), textPlain, forbidden403, ok200)
+--
+--     myHandler :: ProxyRequest NoAuthorizer -> ProxyResponse
+--     myHandler ProxyRequest { httpMethod = \"GET\", path = "/say_hello" } =
+--         ProxyResponse
+--         {   status = ok200
+--         ,   body = textPlain \"Hello\"
+--         ,   headers = mempty
+--         ,   multiValueHeaders = mempty
+--         }
+--     myHandler _ =
+--         ProxyResponse
+--         {   status = forbidden403
+--         ,   body = textPlain \"Forbidden\"
+--         ,   headers = mempty
+--         ,   multiValueHeaders = mempty
+--         }
+--
+--     main :: IO ()
+--     main = pureRuntime myHandler
+-- @
+data ProxyRequest a = ProxyRequest
+    { path                            :: Text
+    , headers                         :: HashMap (CI Text) Text
+    , multiValueHeaders               :: HashMap (CI Text) [Text]
+    , pathParameters                  :: HashMap Text Text
+    , stageVariables                  :: HashMap Text Text
+    , requestContext                  :: RequestContext a
+    , resource                        :: Text
+    , httpMethod                      :: Text
+    , queryStringParameters           :: HashMap Text Text
+    , multiValueQueryStringParameters :: HashMap Text [Text]
+    , body                            :: ByteString
+    } deriving (Generic)
+
+toCIHashMap :: HashMap Text a -> HashMap (CI Text) a
+toCIHashMap = foldrWithKey (insert . mk) mempty
+
+toByteString :: Bool -> TL.Text -> ByteString
+toByteString isBase64Encoded =
+    if isBase64Encoded
+        then decodeLenient . TLE.encodeUtf8
+        else TLE.encodeUtf8
+
+-- | For ignoring API Gateway custom authorizer values
+type NoAuthorizer = Value
+
+-- | For ensuring that there were no API Gateway custom authorizer values (this
+-- is not likely to be useful, you probably want 'NoAuthorizer')
+type StrictlyNoAuthorizer = ()
+
+instance FromJSON a => FromJSON (ProxyRequest a) where
+    parseJSON (Object v) =
+        ProxyRequest <$> v .: "path" <*>
+        (toCIHashMap <$> (v .:? "headers" .!= mempty)) <*>
```

Comment:
I'm slightly surprised there's no constructor that implicitly provides `mempty` :thinking:

Reply:
That’s actually a really good point... I wonder if I missed one (or if PRing one in even makes sense).

Reply:
I skimmed the `aeson` docs but couldn't find anything that required a monoid instance :/ could be worth contributing a new operator/fn?

Reply:
Given:
```
Prelude Data.Foldable> :t fold @Maybe
fold @Maybe :: Monoid m => Maybe m -> m
```
Does `toCIHashMap . fold <$> (v .:? "headers)` do what you want?